feat: improve `screencast` options

danielbayley opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. -->

**What kind of change does this PR introduce?**

<!-- E.g. a bugfix, feature, refactoring, build related change, etc… -->

Adds missing options to `screencast`/`ffmpeg`, and addresses undocumented existing options.

Also fixes multiple bugs in the existing code (see `commit`s…)

Closes #12735

**Did you add tests for your changes?**

Yes

**If relevant, did you update the documentation?**

Yes

**Summary**

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

Adds missing options to `screencast`/`ffmpeg` to better control output, particularly GIFs, along with additional formats. This can avoid having to run `ffmpeg` a second time, only to tweak the final desired output—potentially drastically decreasing, for example, CI run time.

**Does this PR introduce a breaking change?**

<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

No

**Other information**

See also:

#11084

_[High quality GIF with FFmpeg](https://blog.pkh.me/p/21-high-quality-gif-with-ffmpeg.html)_

_[A guide to upscaling or downscaling video with FFmpeg](https://write.corbpie.com/a-guide-to-upscaling-or-downscaling-video-with-ffmpeg)_

_[Elements to Keep in Mind When Optimizing GIFs](https://gifyard.com/elements-for-optimizing-gifs)_

https://github.com/sindresorhus/Gifski/issues/161#issuecomment-552681700

[FFmpeg Filters Documentation](https://ffmpeg.org/ffmpeg-filters.html#palettegen)

Lightning00Blade commented on packages/puppeteer-core/src/node/ScreenRecorder.ts:
```diff
@@ -174,7 +199,14 @@ export class ScreenRecorder extends PassThrough {
+  #getFormatArgs(
+    format: 'webm' | 'gif' | 'mp4',
+    fps: number | 'fps=source_fps',
+    loop: number | typeof Infinity | boolean,
```
Why do we need `typeof Infinity`, 
`const test: number = Infinity;` work in TypeScript. We should allow only `false` and not `boolean`, as having loop true does not make sense without a number

danielbayley replied:
> Why do we need `typeof Infinity`

Suppose I just wanted to emphasize `Infinity` as a nice, semantic option. But simply `loop?: number | Infinity;` is a type error (see related [issue](https://github.com/microsoft/TypeScript/issues/32277)…)

> We should allow only `false` and not `boolean`, as having loop true does not make sense without a number

`loop = true` implies an infinite loop to me, but I guess it's subjective. I went with the below suggestion anyway…

> make this only a number, and it will not be looped if its undefined.

Lightning00Blade commented on packages/puppeteer-core/src/node/ScreenRecorder.ts:
```diff
@@ -40,7 +40,12 @@ const debugFfmpeg = debug('puppeteer:ffmpeg');
-  format?: 'gif' | 'webm';
+  format?: 'gif' | 'webm' | 'mp4';
+  fps?: number;
+  loop?: number | typeof Infinity | boolean;
```
Can we make this only a number, and it will not be looped if its undefined.

Lightning00Blade commented on packages/puppeteer-core/src/node/ScreenRecorder.ts:
```diff
@@ -102,9 +128,10 @@ export class ScreenRecorder extends PassThrough {
+        ['-framerate', `${fps}`],
         // Specifies the encoding and format we are using.
-        this.#getFormatArgs(format ?? 'webm'),
+        // FIXME https://sentry.io/answers/node-js-no-overload-matches-this-call-error-with-typescript
```
What does this `FIXME` refer to? Is it still needed?

danielbayley replied:
Not sure how that one slipped through my version control. There seems to be a bug in _GitKraken_ that sometimes sneakily includes extra adjacent lines…

Lightning00Blade commented on packages/puppeteer-core/src/node/ScreenRecorder.ts:
```diff
@@ -183,21 +215,39 @@ export class ScreenRecorder extends PassThrough {
         ].flat();
       case 'gif':
+        fps = DEFAULT_FPS === fps ? 20 : 'fps=source_fps';
+        if (!Number.isInteger(loop) || loop === 0) loop = ~~!loop * -1;
```
Can't we just set it to the correct default where it won't loop? Rather then this evaluation?

Lightning00Blade commented on packages/puppeteer-core/src/node/ScreenRecorder.ts:
```diff
@@ -183,21 +215,39 @@ export class ScreenRecorder extends PassThrough {
+          // Sets the quality. Lower the better.
+          ['-crf', `${quality}`],
+          // Sets the quality and how efficient the compression will be.
+          ['-deadline', 'realtime', '-cpu-used', '8'],
```
What will happen if the Computer has less then 8 CPU cores?

danielbayley replied:
@Lightning00Blade Presumably, it will explode! Seems to be another oversight in the existing code. Fixed now…
~~~ js
import os from 'node:os'
// …
, '-cpu-used', `${os.cpus().length / 2}`],
~~~

Lightning00Blade commented on packages/puppeteer-core/src/api/Page.ts:
```diff
@@ -345,6 +345,48 @@ export interface ScreencastOptions {
+   * @defaultValue `30` (`20` for GIF)
+   */
+  fps?: number;
+  /**
```
I think some comments are not attached to the correct value, can you please fix?

danielbayley committed:
fix: refactor `screencast` filters

Multiple separate (`-vf`) filters do not work as intended, with each instead overriding the previously specified filters. Therefore, the GIF-specific filters had no effect.

danielbayley committed:
chore: improve `screencast` scaling algorithm

danielbayley committed:
fix: dynamically calculate `screencast` CPU cores

danielbayley committed:
docs: fix missing `screencast` API types

danielbayley committed:
refactor: `screencast` file format type

danielbayley committed:
feat: improve `screencast` API options

danielbayley committed:
docs: add new `screencast` API options

danielbayley committed:
test: new `screencast` API options

danielbayley committed:
chore: add `screencast` GIF dither algorithm

danielbayley committed:
feat: add `screencast` mp4 option

danielbayley committed:
docs: add `screencast` mp4 option

danielbayley committed:
test: add `screencast` mp4 option

danielbayley committed:
fix: `lint` formatting

danielbayley committed:
docs: generated from `npm run docs`

Lightning00Blade commented on packages/puppeteer-core/src/node/ScreenRecorder.ts:
```diff
@@ -56,25 +62,73 @@ export class ScreenRecorder extends PassThrough {
+      colors,
+    );
+    const vf = formatArgs.indexOf('-vf');
+    if (~vf) {
```
Let's just check `if (vf !== -1)` it easier to understand.
Also in the next line if we push a empty string won't we end up with `<previous-filter>,,`
Which could break the command line in theory?

danielbayley replied:
> Also in the next line if we push a empty string won't we end up with <previous-filter>,,
Which could break the command line in theory?

@Lightning00Blade Well I tested all 3 output formats, and nothing broke. But I agree it's not ideal… Also, this was purely to stop TypeScript complaining about possible `undefined`!

Lightning00Blade commented on packages/puppeteer-core/src/api/Page.ts:
```diff
@@ -345,6 +353,42 @@ export interface ScreencastOptions {
+  /**
+   * Specifies the number of times to loop playback, from `0` to `Infinity`.
+   *
+   * @defaultValue `0`
```
I think we want a little better explanation here, 
As if undefined is provided no loop;
If 0 or Infinity is provided - loop Infinity
And default is undefined.

danielbayley replied:
> If 0 or Infinity is provided - loop Infinity

@Lightning00Blade Actually, I intentionally set `0` to prevent looping:

https://github.com/puppeteer/puppeteer/blob/505d75e60b2ab96c8a155b87182d7bf483607fb9/packages/puppeteer-core/src/node/ScreenRecorder.ts#L91

as this is a very confusing aspect of the underlying `ffmpeg` API. This simplifies it to be `0` for no loop, `1` to loop once, `2`, twice, etc. up to `Infinity`… So actually `0` is the same as `undefined` (the default) here. Reworded it anyway, to clarify…

Lightning00Blade replied:
I think it looks nice with the latest implementation. @OrKoN WDYT ?

danielbayley committed:
refactor: `screencast` filter check

danielbayley committed:
docs: reword `screencast` GIF `loop` option

danielbayley committed:
fix: `screencast` delay check

Lightning00Blade commented on packages/puppeteer-core/src/node/ScreenRecorder.ts:
```diff
@@ -56,25 +62,73 @@ export class ScreenRecorder extends PassThrough {
 
+    format ??= 'webm';
+    fps ??= DEFAULT_FPS;
+    loop ||= -1;
```
Can we add comment why we did this, I already forgot when reading the from the start:
Maps 0 to -1 as FFMPEG maps 0 to Infinity.

Lightning00Blade commented on packages/puppeteer-core/src/node/ScreenRecorder.ts:
```diff
@@ -174,30 +220,59 @@ export class ScreenRecorder extends PassThrough {
+        }
+        if (delay !== -1) {
+          delay /= 10;
+        } // ms to cs
```
Can you move this comment above the the delay /=10;

danielbayley replied:
> Can you move this comment above the the delay /=10;

Sure. `npm run format` did it that way… Probably a bug in `prettier`?

Lightning00Blade replied:
Yeah, I think it's an edge case with adding brackets when no are provided for if statement.

danielbayley replied:
I personally prefer without `{}`, wherever possible anyway :)

Also, _[Biome](https://biomejs.dev/formatter)_ > [eslint-stylistic `--fix`](https://antfu.me/posts/why-not-prettier) > Prettier

Lightning00Blade reviewed (commented):
From code perspective all looks good to me.
I will double check the FFMPEG integration and options and then we can merge.

danielbayley committed:
fix: refactor `screencast` comments

Lightning00Blade commented on packages/puppeteer-core/src/node/ScreenRecorder.ts:
```diff
@@ -174,30 +221,60 @@ export class ScreenRecorder extends PassThrough {
+      // Sets the quality. Lower the better.
+      ['-crf', `${quality}`],
+      // Sets the quality and how efficient the compression will be.
+      ['-deadline', 'realtime', '-cpu-used', `${os.cpus().length / 2}`],
```
Funny bug its max value here is 8 - https://trac.ffmpeg.org/wiki/Encode/VP9#CPUUtilizationSpeed
```suggestion
      ['-deadline', 'realtime', '-cpu-used', `${Math.min(os.cpus().length / 2, 8)}`],
```

danielbayley committed:
fix: `screencast` CPU cores limit

Co-authored-by: Nikolay Vitkov <34244704+Lightning00Blade@users.noreply.github.com>

danielbayley committed:
fix: remove `screencast` loop option test

danielbayley committed:
fix: `screencast` mp4 encoding

Lightning00Blade commented on packages/puppeteer-core/src/api/Page.ts:
```diff
@@ -318,11 +318,19 @@ export interface ScreenshotOptions {
  */
+export type FileFormat = 'gif' | 'webm' | 'mp4';
+
 export interface ScreencastOptions {
```
There is a missing comment with TsDoc, please copy it to this interface from the FileFormat.

danielbayley committed:
fix: `screencast` docs API file format type

Lightning00Blade reviewed (approved):
LGTM, thanks again for the PR.


PS: I will deal with the rebasing and merging.
